### PR TITLE
containerd support

### PR DESCRIPTION
**Pull request by wenchajun**
Thank you for your comments.I have revised and resubmitted the PR. Please review it .
Install the version without the plug-in

```
helm install fluentbit-operator  --create-namespace -n kubesphere-logging-system chart/fluentbit-operator/charts/k8s/
```

Install the version with the plug-in

```
helm install  --create-namespace -n kubesphere-logging-system chart/fluentbit-operator/ --generate-name
```
You should change the `Parser` on the input from  `docker` to `cri` . And you can  set  `values.input.cri`  to `true` :

```
helm install fluentbit-operator  --create-namespace -n kubesphere-logging-system chart/fluentbit-operator/
```

### CODE REVIEW

**Comment by wenchajun** [comment]
I have updated some files. Now collecting containerd logs just need to change the `Parser` on the input from  `docker` to `cri` . No additional installation is required for users.

File: chart/fluentbit-operator/charts/k8s/values.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Default values for fluentbit-operator.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+fluentbitOperatorImage:
```
Comment (benjaminhuo) [inline chart/fluentbit-operator/charts/k8s/values.yaml:4]: This should be consistent with below fluentbit section:
```
operator:
  image: "kubesphere/fluentbit-operator"
  tag: "v0.8.0"
```

File: charts/logging-stack/charts/fluentbit-operator/values.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# Default values for fluentbit-operator.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+fluentbitOperatorImage:
```
Comment (benjaminhuo) [inline charts/logging-stack/charts/fluentbit-operator/values.yaml:4]: fluentbitOperatorImage => operator

File: charts/logging-stack/values.yaml [hunk 3]
```diff
@@ -75,6 +43,9 @@ resources:
 input:
   tail:
     memBufLimit: 5MB
+#Set this to cri if you want to collect containerd or cri-o information
```
Comment (benjaminhuo) [inline charts/logging-stack/values.yaml:46]: =>
# Set this to cri if you want to collect CRI format logs for containerd or CRI-O.

**Review by wanjunlei — CHANGES_REQUESTED** [review]
Please update the catalog of README as well.

**Review by wanjunlei — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -76,14 +84,30 @@ kubectl apply -f https://raw.githubusercontent.com/kubesphere/fluentbit-operator
...
 The quick start instructs you to deploy fluent bit with `dummy` as input and `stdout` as output, which is equivalent to execute the binary with `fluent-bit -i dummy -o stdout`.
 
+##### Deploy Quick Start with YAML
+
 ```shell
 kubectl apply -f https://raw.githubusercontent.com/kubesphere/fluentbit-operator/master/manifests/quick-start/quick-start.yaml
 ```
 
+##### Deploy Quick Start with Helm chart
+
```
Comment (benjaminhuo) [inline README.md:106]: We needn't add a helm switch for quick start

File: README.md [hunk 5]
```diff
@@ -108,6 +132,8 @@ This guide provisions a logging pipeline including the Fluent Bit DaemonSet and
 
 > Note that you need a running Elasticsearch v5+ cluster to receive log data before start. **Remember to adjust [output-elasticsearch.yaml](manifests/logging-stack/output-elasticsearch.yaml) to your own es setup**. Kafka and Fluentd outputs are optional and are turned off by default.
 
+##### Deploy logging-stack with YAML
```
Comment (benjaminhuo) [inline README.md:135]: => 
Deploy the Kubernetes logging stack with YAML

File: README.md [hunk 6]
```diff
@@ -116,12 +142,27 @@ kubectl apply -f manifests/logging-stack
 # kubectl kustomize manifests/logging-stack/ | kubectl apply -f -
 ```
 
+##### Deploy logging-stack with Helm chart
```
Comment (benjaminhuo) [inline README.md:145]: =>
Deploy the Kubernetes logging stack with Helm

File: README.md [hunk 7]
```diff
@@ -76,14 +84,30 @@ kubectl apply -f https://raw.githubusercontent.com/kubesphere/fluentbit-operator
 # kubectl kustomize manifests/setup/ | kubectl apply -f -
 ```
 
+##### Deploy Fluent Bit Operator with Helm chart
```
Comment (benjaminhuo) [inline README.md:87]: =>
Deploy Fluent Bit Operator with Helm

File: charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        hostPath:
+          path: /usr/local/bin/crictl
+      initContainers:
+      - name: setenv
+        image: {{ .Values.operator.initcontainer.image }}:{{ .Values.operator.initcontainer.tag }}
+        command:
+        - /bin/sh
+        - '-c'
+        - set -ex;
+          echo DOCKER_ROOT_DIR=$(crictl info --output go-template --template '{{`{{ .config.containerdRootDir }}`}}') > /fluentbit-operator/fluent-bit.env
```
Comment (wanjunlei) [inline charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml:57]: Maybe `LOG_ROOT_DIR` or `CONTAINER_ROOT_DIR` is more suitable?

Comment (wenchajun) [reply charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml:57]: Just to match this https://github.com/fluent/fluentbit-operator/blob/master/cmd/manager/main.go#L59

Comment (wenchajun) [reply charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml]: I have finished the modification, please review it

**Comment by benjaminhuo** [comment]
@wenchajun the clusterrole and clusterrolebinding has been changed to role and rolebinding by this PR https://github.com/fluent/fluentbit-operator/pull/117 , we should adjust the helm accordingly

**Comment by wenchajun** [comment]
> @wenchajun the clusterrole and clusterrolebinding has been changed to role and rolebinding by this PR #117 , we should adjust the helm accordingly

thanks, I have fixed it.

File: charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        - '-c'
+        - set -ex;
+          echo CONTAINER_ROOT_DIR=$(crictl info --output go-template --template '{{`{{ .config.containerdRootDir }}`}}') > /fluentbit-operator/fluent-bit.env
+        volumeMounts:
+        - name: env
+          mountPath: /fluentbit-operator
+        - name: containerdsock
+          readOnly: true
+          mountPath: /var/run/containerd/containerd.sock
+        - name: containerdsock
```
Comment (nicktate) [inline charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml:64]: duplicate volume mount

File: README.md [hunk 10]
```diff
@@ -76,6 +82,30 @@ kubectl apply -f https://raw.githubusercontent.com/kubesphere/fluentbit-operator
 # kubectl kustomize manifests/setup/ | kubectl apply -f -
 ```
 
+##### Deploy Fluent Bit Operator with Helm 
+
+> Note: For the Helm-based installation you need Helm v3.2.1 or later.
+
+Fluent Bit Operator supports `docker` as well as `containerd` and `CRI-O`. `containerd` and `CRI-O` use the `CRI Log` format which is slightly different and requires additional parsing to parse JSON application logs. You should set different `containerRuntime` depending on  your own container runtime.
```
Comment (benjaminhuo) [inline README.md:89]: depending on  your own container runtime.
=>
depending on your container runtime.

File: README.md [hunk 11]
```diff
@@ -76,6 +82,30 @@ kubectl apply -f https://raw.githubusercontent.com/kubesphere/fluentbit-operator
...
+
+If your own container runtime is `docker`
+
+```shell
+helm install fluentbit-operator  --create-namespace -n kubesphere-logging-system charts/fluentbit-operator/--set containerRuntime=docker
+```
+
+If your own container runtime is `containerd`
```
Comment (benjaminhuo) [inline README.md:97]: If your own container runtime
=>
If your container runtime

File: charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+          hostPath:
+            path: /var/run/crio/crio.sock
+      initContainers:
+      - name: setenv
+        image: {{ .Values.operator.initcontainer.image }}:{{ .Values.operator.initcontainer.tag }}
+        command:
+        - /bin/sh
+        - '-c'
+        - set -ex;
+          echo CONTAINER_ROOT_DIR=/var/log > /fluentbit-operator/fluent-bit.env
```
Comment (benjaminhuo) [inline charts/fluentbit-operator/templates/fluentbit-operator-deployment.yaml:81]: If the logpath for crio is hardcoded here, we should add a helm param logPath for this

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**